ngi_reports/utils/entities.py

```diff
@@ -0,0 +1,453 @@
+                    if fc['db'] == 'x_flowcells':
+                        sample = stat['Sample']
+                        barcode = stat['Barcode sequence']
+                        qval_key, base_key = ('% >= Q30bases', 'PF Clusters')
+
+                    else:
+                        sample = stat['Sample ID']
+                        barcode = stat['Index']
+                        qval_key, base_key = ('% of >= Q30 Bases (PF)', '# Reads')
+
+                    if kwargs.get('samples', []) and sample not in kwargs.get('samples', []):
+                        continue
+
+                    try:
+                        r_idx = '{}_{}_{}'.format(lane, fcObj.name, barcode)
+                        r_len_list = [x['NumCycles'] for x in fcObj.run_setup if x['IsIndexedRead'] == 'N']
+                        r_len_list = [int(x) for x in r_len_list]
+                        r_num = len(r_len_list)
+                        qval = float(stat.get(qval_key))
+                        pfrd = int(stat.get(base_key).replace(',',''))
+                        pfrd = pfrd/2 if fc['db'] == 'flowcell' else pfrd
+                        base = pfrd * sum(r_len_list)
+                        sample_qval[sample][r_idx] = {'qval': qval, 'reads': pfrd, 'bases': base}
+
+                    except (TypeError, ValueError, AttributeError) as e:
+                        log.warn('Something went wrong while fetching Q30 for sample {} with barcode {} in FC {} at lane {}'.format(sample, barcode, fcObj.name, lane))
+                        pass
+                    ## collect lanes of interest to proceed later
+                    fc_lane_summary = fc_details.get('lims_data', {}).get('run_summary', {})
+                    if lane not in fcObj.lanes:
+                        laneObj = Lane()
+                        lane_sum = fc_lane_summary.get(lane, fc_lane_summary.get('A',{}))
+                        laneObj.id = lane
+                        laneObj.set_lane_info('cluster', 'Reads PF (M)' if 'NovaSeq' in fcObj.type or 'NextSeq' in fcObj.type else 'Clusters PF', lane_sum,
+                                                    str(r_num), False if 'NovaSeq' in fcObj.type or 'NextSeq' in fcObj.type else True)
+                        laneObj.set_lane_info('avg_qval', '% Bases >=Q30', lane_sum, str(r_num))
+                        laneObj.set_lane_info('fc_phix', '% Error Rate', lane_sum, str(r_num))
+                        if kwargs.get('fc_phix',{}).get(fcObj.name, {}):
+                            laneObj.phix = kwargs.get('fc_phix').get(fcObj.name).get(lane)
+
+                        fcObj.lanes[lane] = laneObj
+
+                        ## Check if the above created lane object has all needed info
+                        for k,v in vars(laneObj).items():
+                            if not v:
+                                log.warn('Could not fetch {} for FC {} at lane {}'.format(k, fcObj.name, lane))
+                except KeyError:
```

Comment:
GAH! Any KeyError within the previous 40 lines of code is just fine? Super easy to miss some more serious errors this way.

Reply:
But I guess it just affects the quality values anyway...

Reply:
Changed it somewhat based on what I think it was originally for.